File: server/routes/savedViews.js
const express = require('express');
const router = express.Router();
const db = require('../db/database');
const { v4: uuidv4 } = require('uuid');

// Helper to generate timestamp
const getTimestamp = () => new Date().toISOString();

// GET /api/saved-views - List all saved views (optional ?project_id=&view_type= filters)
router.get('/', (req, res) => {
  try {
    const { project_id, view_type } = req.query;
    
    let query = `
      SELECT sv.*, p.name as project_name
      FROM saved_views sv
      LEFT JOIN projects p ON sv.project_id = p.id
      WHERE 1=1
    `;
    let params = [];
    
    if (project_id) {
      // Get project-specific views AND global views (project_id IS NULL)
      query += ` AND (sv.project_id = ? OR sv.project_id IS NULL)`;
      params.push(project_id);
    }
    
    if (view_type) {
      query += ` AND sv.view_type = ?`;
      params.push(view_type);
    }
    
    query += ` ORDER BY sv.name ASC`;
    
    const views = db.prepare(query).all(...params);
    
    // Parse filters JSON and convert is_default to boolean
    const parsedViews = views.map(view => ({
      ...view,
      filters: view.filters ? JSON.parse(view.filters) : {},
      is_default: !!view.is_default
    }));
    
    res.json({ success: true, data: parsedViews });
  } catch (error) {
    console.error('Error fetching saved views:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'FETCH_ERROR', message: error.message } 
    });
  }
});

// GET /api/saved-views/:id - Get single saved view
router.get('/:id', (req, res) => {
  try {
    const { id } = req.params;
    
    const view = db.prepare(`
      SELECT sv.*, p.name as project_name
      FROM saved_views sv
      LEFT JOIN projects p ON sv.project_id = p.id
      WHERE sv.id = ?
    `).get(id);
    
    if (!view) {
      return res.status(404).json({ 
        success: false, 
        error: { code: 'NOT_FOUND', message: 'Saved view not found' } 
      });
    }
    
    // Parse filters JSON
    view.filters = view.filters ? JSON.parse(view.filters) : {};
    view.is_default = !!view.is_default;
    
    res.json({ success: true, data: view });
  } catch (error) {
    console.error('Error fetching saved view:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'FETCH_ERROR', message: error.message } 
    });
  }
});

// POST /api/saved-views - Create saved view
router.post('/', (req, res) => {
  try {
    const { name, view_type, project_id, filters, sort_by, sort_order, is_default } = req.body;
    
    // Validation
    if (!name || !view_type) {
      return res.status(400).json({ 
        success: false, 
        error: { code: 'VALIDATION_ERROR', message: 'Name and view_type are required' } 
      });
    }
    
    const validViewTypes = ['list', 'kanban', 'calendar', 'timeline'];
    if (!validViewTypes.includes(view_type)) {
      return res.status(400).json({ 
        success: false, 
        error: { code: 'VALIDATION_ERROR', message: `view_type must be one of: ${validViewTypes.join(', ')}` } 
      });
    }
    
    const id = uuidv4();
    const now = getTimestamp();
    const filtersJson = JSON.stringify(filters || {});
    
    // If this is set as default, unset any existing default for the same project and view_type
    if (is_default) {
      const unsetQuery = project_id
        ? `UPDATE saved_views SET is_default = 0 WHERE project_id = ? AND view_type = ?`
        : `UPDATE saved_views SET is_default = 0 WHERE project_id IS NULL AND view_type = ?`;
      
      if (project_id) {
        db.prepare(unsetQuery).run(project_id, view_type);
      } else {
        db.prepare(unsetQuery).run(view_type);
      }
    }
    
    const stmt = db.prepare(`
      INSERT INTO saved_views (id, name, view_type, project_id, filters, sort_by, sort_order, is_default, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);
    
    stmt.run(
      id, 
      name, 
      view_type, 
      project_id || null, 
      filtersJson, 
      sort_by || null,
      sort_order || 'asc',
      is_default ? 1 : 0,
      now, 
      now
    );
    
    const newView = db.prepare(`
      SELECT sv.*, p.name as project_name
      FROM saved_views sv
      LEFT JOIN projects p ON sv.project_id = p.id
      WHERE sv.id = ?
    `).get(id);
    
    newView.filters = newView.filters ? JSON.parse(newView.filters) : {};
    newView.is_default = !!newView.is_default;
    
    res.status(201).json({ success: true, data: newView });
  } catch (error) {
    console.error('Error creating saved view:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'CREATE_ERROR', message: error.message } 
    });
  }
});

// PUT /api/saved-views/:id - Update saved view
router.put('/:id', (req, res) => {
  try {
    const { id } = req.params;
    const { name, view_type, project_id, filters, sort_by, sort_order, is_default } = req.body;
    
    // Check if view exists
    const existingView = db.prepare('SELECT * FROM saved_views WHERE id = ?').get(id);
    if (!existingView) {
      return res.status(404).json({ 
        success: false, 
        error: { code: 'NOT_FOUND', message: 'Saved view not found' } 
      });
    }
    
    // Validate view_type if provided
    if (view_type) {
      const validViewTypes = ['list', 'kanban', 'calendar', 'timeline'];
      if (!validViewTypes.includes(view_type)) {
        return res.status(400).json({ 
          success: false, 
          error: { code: 'VALIDATION_ERROR', message: `view_type must be one of: ${validViewTypes.join(', ')}` } 
        });
      }
    }
    
    const now = getTimestamp();
    const filtersJson = filters !== undefined ? JSON.stringify(filters) : existingView.filters;
    const effectiveViewType = view_type || existingView.view_type;
    const effectiveProjectId = project_id !== undefined ? (project_id || null) : existingView.project_id;
    
    // If this is being set as default, unset any existing default for the same project and view_type
    if (is_default) {
      const unsetQuery = effectiveProjectId
        ? `UPDATE saved_views SET is_default = 0 WHERE project_id = ? AND view_type = ? AND id != ?`
        : `UPDATE saved_views SET is_default = 0 WHERE project_id IS NULL AND view_type = ? AND id != ?`;
      
      if (effectiveProjectId) {
        db.prepare(unsetQuery).run(effectiveProjectId, effectiveViewType, id);
      } else {
        db.prepare(unsetQuery).run(effectiveViewType, id);
      }
    }
    
    const stmt = db.prepare(`
      UPDATE saved_views 
      SET name = ?, view_type = ?, project_id = ?, filters = ?, sort_by = ?, sort_order = ?, is_default = ?, updated_at = ?
      WHERE id = ?
    `);
    
    stmt.run(
      name ?? existingView.name,
      effectiveViewType,
      effectiveProjectId,
      filtersJson,
      sort_by !== undefined ? sort_by : existingView.sort_by,
      sort_order !== undefined ? sort_order : existingView.sort_order,
      is_default !== undefined ? (is_default ? 1 : 0) : existingView.is_default,
      now,
      id
    );
    
    const updatedView = db.prepare(`
      SELECT sv.*, p.name as project_name
      FROM saved_views sv
      LEFT JOIN projects p ON sv.project_id = p.id
      WHERE sv.id = ?
    `).get(id);
    
    updatedView.filters = updatedView.filters ? JSON.parse(updatedView.filters) : {};
    updatedView.is_default = !!updatedView.is_default;
    
    res.json({ success: true, data: updatedView });
  } catch (error) {
    console.error('Error updating saved view:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'UPDATE_ERROR', message: error.message } 
    });
  }
});

// DELETE /api/saved-views/:id - Delete saved view
router.delete('/:id', (req, res) => {
  try {
    const { id } = req.params;
    
    const existingView = db.prepare('SELECT * FROM saved_views WHERE id = ?').get(id);
    if (!existingView) {
      return res.status(404).json({ 
        success: false, 
        error: { code: 'NOT_FOUND', message: 'Saved view not found' } 
      });
    }
    
    db.prepare('DELETE FROM saved_views WHERE id = ?').run(id);
    
    res.json({ success: true, data: { id, message: 'Saved view deleted successfully' } });
  } catch (error) {
    console.error('Error deleting saved view:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'DELETE_ERROR', message: error.message } 
    });
  }
});

// PUT /api/saved-views/:id/set-default - Set as default view for project/type
router.put('/:id/set-default', (req, res) => {
  try {
    const { id } = req.params;
    
    // Check if view exists
    const existingView = db.prepare('SELECT * FROM saved_views WHERE id = ?').get(id);
    if (!existingView) {
      return res.status(404).json({ 
        success: false, 
        error: { code: 'NOT_FOUND', message: 'Saved view not found' } 
      });
    }
    
    const now = getTimestamp();
    
    // Unset any existing default for the same project and view_type
    const unsetQuery = existingView.project_id
      ? `UPDATE saved_views SET is_default = 0 WHERE project_id = ? AND view_type = ?`
      : `UPDATE saved_views SET is_default = 0 WHERE project_id IS NULL AND view_type = ?`;
    
    if (existingView.project_id) {
      db.prepare(unsetQuery).run(existingView.project_id, existingView.view_type);
    } else {
      db.prepare(unsetQuery).run(existingView.view_type);
    }
    
    // Set this view as default
    db.prepare(`UPDATE saved_views SET is_default = 1, updated_at = ? WHERE id = ?`).run(now, id);
    
    const updatedView = db.prepare(`
      SELECT sv.*, p.name as project_name
      FROM saved_views sv
      LEFT JOIN projects p ON sv.project_id = p.id
      WHERE sv.id = ?
    `).get(id);
    
    updatedView.filters = updatedView.filters ? JSON.parse(updatedView.filters) : {};
    updatedView.is_default = !!updatedView.is_default;
    
    res.json({ success: true, data: updatedView });
  } catch (error) {
    console.error('Error setting default view:', error);
    res.status(500).json({ 
      success: false, 
      error: { code: 'UPDATE_ERROR', message: error.message } 
    });
  }
});

module.exports = router;

File: server/internal/handlers/saved_views.go
package handlers

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"time"

	"github.com/celestask/server/internal/db"
	"github.com/celestask/server/internal/middleware"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// Valid view types
var validViewTypes = []string{"list", "kanban", "calendar", "timeline"}

// SavedView represents a saved view in the database
type SavedView struct {
	ID          string         `json:"id"`
	Name        string         `json:"name"`
	ViewType    string         `json:"view_type"`
	ProjectID   sql.NullInt64  `json:"project_id"`
	Filters     string         `json:"filters"`
	SortBy      sql.NullString `json:"sort_by"`
	SortOrder   sql.NullString `json:"sort_order"`
	IsDefault   int            `json:"is_default"`
	CreatedAt   string         `json:"created_at"`
	UpdatedAt   string         `json:"updated_at"`
	ProjectName sql.NullString `json:"project_name"`
}

// SavedViewResponse is the API response format
type SavedViewResponse struct {
	ID          string                 `json:"id"`
	Name        string                 `json:"name"`
	ViewType    string                 `json:"view_type"`
	ProjectID   *int64                 `json:"project_id"`
	ProjectName *string                `json:"project_name"`
	Filters     map[string]interface{} `json:"filters"`
	SortBy      *string                `json:"sort_by"`
	SortOrder   string                 `json:"sort_order"`
	IsDefault   bool                   `json:"is_default"`
	CreatedAt   string                 `json:"created_at"`
	UpdatedAt   string                 `json:"updated_at"`
}

// CreateSavedViewRequest is the request body for creating a saved view
type CreateSavedViewRequest struct {
	Name      string                 `json:"name" binding:"required"`
	ViewType  string                 `json:"view_type" binding:"required"`
	ProjectID *int64                 `json:"project_id"`
	Filters   map[string]interface{} `json:"filters"`
	SortBy    *string                `json:"sort_by"`
	SortOrder *string                `json:"sort_order"`
	IsDefault *bool                  `json:"is_default"`
}

// UpdateSavedViewRequest is the request body for updating a saved view
type UpdateSavedViewRequest struct {
	Name      *string                `json:"name"`
	ViewType  *string                `json:"view_type"`
	ProjectID *int64                 `json:"project_id"`
	Filters   map[string]interface{} `json:"filters"`
	SortBy    *string                `json:"sort_by"`
	SortOrder *string                `json:"sort_order"`
	IsDefault *bool                  `json:"is_default"`
}

// isValidViewType checks if the view type is valid
func isValidViewType(viewType string) bool {
	for _, vt := range validViewTypes {
		if vt == viewType {
			return true
		}
	}
	return false
}

// parseSavedView converts a database row to API response
func parseSavedView(row *SavedView) SavedViewResponse {
	response := SavedViewResponse{
		ID:        row.ID,
		Name:      row.Name,
		ViewType:  row.ViewType,
		Filters:   make(map[string]interface{}),
		SortOrder: "asc",
		IsDefault: row.IsDefault == 1,
		CreatedAt: row.CreatedAt,
		UpdatedAt: row.UpdatedAt,
	}

	if row.ProjectID.Valid {
		response.ProjectID = &row.ProjectID.Int64
	}

	if row.ProjectName.Valid {
		response.ProjectName = &row.ProjectName.String
	}

	if row.Filters != "" {
		json.Unmarshal([]byte(row.Filters), &response.Filters)
	}

	if row.SortBy.Valid {
		response.SortBy = &row.SortBy.String
	}

	if row.SortOrder.Valid {
		response.SortOrder = row.SortOrder.String
	}

	return response
}

// GetSavedViews retrieves all saved views, optionally filtered by project_id and view_type
func GetSavedViews(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	projectID := c.Query("project_id")
	viewType := c.Query("view_type")

	query := `
		SELECT sv.*, p.name as project_name
		FROM saved_views sv
		LEFT JOIN projects p ON sv.project_id = p.id
		WHERE 1=1
	`
	args := []interface{}{}

	if projectID != "" {
		// Get project-specific views AND global views (project_id IS NULL)
		query += " AND (sv.project_id = ? OR sv.project_id IS NULL)"
		args = append(args, projectID)
	}

	if viewType != "" {
		query += " AND sv.view_type = ?"
		args = append(args, viewType)
	}

	query += " ORDER BY sv.name ASC"

	rows, err := database.Query(query, args...)
	if err != nil {
		panic(middleware.NewFetchError("saved views"))
	}
	defer rows.Close()

	var views []SavedViewResponse
	for rows.Next() {
		var view SavedView
		err := rows.Scan(
			&view.ID,
			&view.Name,
			&view.ViewType,
			&view.ProjectID,
			&view.Filters,
			&view.SortBy,
			&view.SortOrder,
			&view.IsDefault,
			&view.CreatedAt,
			&view.UpdatedAt,
			&view.ProjectName,
		)
		if err != nil {
			panic(middleware.NewFetchError("saved views"))
		}
		parsedView := parseSavedView(&view)
		views = append(views, parsedView)
	}

	if views == nil {
		views = []SavedViewResponse{}
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(views))
}

// GetSavedView retrieves a single saved view by ID
func GetSavedView(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	id := c.Param("id")

	var view SavedView
	err := database.QueryRow(`
		SELECT sv.*, p.name as project_name
		FROM saved_views sv
		LEFT JOIN projects p ON sv.project_id = p.id
		WHERE sv.id = ?
	`, id).Scan(
		&view.ID,
		&view.Name,
		&view.ViewType,
		&view.ProjectID,
		&view.Filters,
		&view.SortBy,
		&view.SortOrder,
		&view.IsDefault,
		&view.CreatedAt,
		&view.UpdatedAt,
		&view.ProjectName,
	)

	if err == sql.ErrNoRows {
		panic(middleware.NewNotFoundError("Saved view"))
	}

	if err != nil {
		panic(middleware.NewFetchError("saved view"))
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(parseSavedView(&view)))
}

// CreateSavedView creates a new saved view
func CreateSavedView(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	var req CreateSavedViewRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		panic(middleware.NewValidationError("Name and view_type are required"))
	}

	// Validate view_type
	if !isValidViewType(req.ViewType) {
		panic(middleware.NewValidationError("view_type must be one of: list, kanban, calendar, timeline"))
	}

	id := uuid.New().String()
	now := time.Now().Format(time.RFC3339)

	filtersJSON := "{}"
	if req.Filters != nil {
		filtersBytes, _ := json.Marshal(req.Filters)
		filtersJSON = string(filtersBytes)
	}

	sortOrder := "asc"
	if req.SortOrder != nil {
		sortOrder = *req.SortOrder
	}

	isDefault := 0
	if req.IsDefault != nil && *req.IsDefault {
		isDefault = 1
		// Unset any existing default for the same project and view_type
		if req.ProjectID != nil {
			database.Exec("UPDATE saved_views SET is_default = 0 WHERE project_id = ? AND view_type = ?", *req.ProjectID, req.ViewType)
		} else {
			database.Exec("UPDATE saved_views SET is_default = 0 WHERE project_id IS NULL AND view_type = ?", req.ViewType)
		}
	}

	var projectID interface{}
	if req.ProjectID != nil {
		projectID = *req.ProjectID
	} else {
		projectID = nil
	}

	_, err := database.Exec(`
		INSERT INTO saved_views (id, name, view_type, project_id, filters, sort_by, sort_order, is_default, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
	`, id, req.Name, req.ViewType, projectID, filtersJSON, req.SortBy, sortOrder, isDefault, now, now)

	if err != nil {
		panic(middleware.NewCreateError("saved view"))
	}

	// Fetch the created view
	var view SavedView
	err = database.QueryRow(`
		SELECT sv.*, p.name as project_name
		FROM saved_views sv
		LEFT JOIN projects p ON sv.project_id = p.id
		WHERE sv.id = ?
	`, id).Scan(
		&view.ID,
		&view.Name,
		&view.ViewType,
		&view.ProjectID,
		&view.Filters,
		&view.SortBy,
		&view.SortOrder,
		&view.IsDefault,
		&view.CreatedAt,
		&view.UpdatedAt,
		&view.ProjectName,
	)

	if err != nil {
		panic(middleware.NewFetchError("saved view"))
	}

	c.JSON(http.StatusCreated, middleware.NewSuccessResponse(parseSavedView(&view)))
}

// UpdateSavedView updates an existing saved view
func UpdateSavedView(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	id := c.Param("id")

	var req UpdateSavedViewRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		panic(middleware.NewValidationError("Invalid request body"))
	}

	// Check if view exists
	var existingView SavedView
	err := database.QueryRow("SELECT * FROM saved_views WHERE id = ?", id).Scan(
		&existingView.ID,
		&existingView.Name,
		&existingView.ViewType,
		&existingView.ProjectID,
		&existingView.Filters,
		&existingView.SortBy,
		&existingView.SortOrder,
		&existingView.IsDefault,
		&existingView.CreatedAt,
		&existingView.UpdatedAt,
	)

	if err == sql.ErrNoRows {
		panic(middleware.NewNotFoundError("Saved view"))
	}

	// Validate view_type if provided
	if req.ViewType != nil && !isValidViewType(*req.ViewType) {
		panic(middleware.NewValidationError("view_type must be one of: list, kanban, calendar, timeline"))
	}

	now := time.Now().Format(time.RFC3339)

	name := existingView.Name
	if req.Name != nil {
		name = *req.Name
	}

	viewType := existingView.ViewType
	if req.ViewType != nil {
		viewType = *req.ViewType
	}

	var projectID interface{}
	if req.ProjectID != nil {
		projectID = *req.ProjectID
	} else {
		projectID = existingView.ProjectID
	}

	filtersJSON := existingView.Filters
	if req.Filters != nil {
		filtersBytes, _ := json.Marshal(req.Filters)
		filtersJSON = string(filtersBytes)
	}

	sortBy := existingView.SortBy.String
	if req.SortBy != nil {
		sortBy = *req.SortBy
	}

	sortOrder := existingView.SortOrder.String
	if req.SortOrder != nil {
		sortOrder = *req.SortOrder
	}

	isDefault := existingView.IsDefault
	if req.IsDefault != nil {
		if *req.IsDefault {
			isDefault = 1
			// Unset any existing default for the same project and view_type
			if projectID != nil {
				database.Exec("UPDATE saved_views SET is_default = 0 WHERE project_id = ? AND view_type = ? AND id != ?", projectID, viewType, id)
			} else {
				database.Exec("UPDATE saved_views SET is_default = 0 WHERE project_id IS NULL AND view_type = ? AND id != ?", viewType, id)
			}
		} else {
			isDefault = 0
		}
	}

	_, err = database.Exec(`
		UPDATE saved_views 
		SET name = ?, view_type = ?, project_id = ?, filters = ?, sort_by = ?, sort_order = ?, is_default = ?, updated_at = ?
		WHERE id = ?
	`, name, viewType, projectID, filtersJSON, sortBy, sortOrder, isDefault, now, id)

	if err != nil {
		panic(middleware.NewUpdateError("saved view"))
	}

	// Fetch the updated view
	var view SavedView
	err = database.QueryRow(`
		SELECT sv.*, p.name as project_name
		FROM saved_views sv
		LEFT JOIN projects p ON sv.project_id = p.id
		WHERE sv.id = ?
	`, id).Scan(
		&view.ID,
		&view.Name,
		&view.ViewType,
		&view.ProjectID,
		&view.Filters,
		&view.SortBy,
		&view.SortOrder,
		&view.IsDefault,
		&view.CreatedAt,
		&view.UpdatedAt,
		&view.ProjectName,
	)

	if err != nil {
		panic(middleware.NewFetchError("saved view"))
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(parseSavedView(&view)))
}

// DeleteSavedView deletes a saved view
func DeleteSavedView(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	id := c.Param("id")

	// Check if view exists
	var existingView SavedView
	err := database.QueryRow("SELECT * FROM saved_views WHERE id = ?", id).Scan(
		&existingView.ID,
		&existingView.Name,
		&existingView.ViewType,
		&existingView.ProjectID,
		&existingView.Filters,
		&existingView.SortBy,
		&existingView.SortOrder,
		&existingView.IsDefault,
		&existingView.CreatedAt,
		&existingView.UpdatedAt,
	)

	if err == sql.ErrNoRows {
		panic(middleware.NewNotFoundError("Saved view"))
	}

	if err != nil {
		panic(middleware.NewFetchError("saved view"))
	}

	_, err = database.Exec("DELETE FROM saved_views WHERE id = ?", id)
	if err != nil {
		panic(middleware.NewDeleteError("saved view"))
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(gin.H{
		"id":      id,
		"message": "Saved view deleted successfully",
	}))
}

// SetDefaultView sets a saved view as the default for its project/type combination
func SetDefaultView(c *gin.Context) {
	database, ok := c.MustGet("database").(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	id := c.Param("id")

	// Check if view exists
	var existingView SavedView
	err := database.QueryRow("SELECT * FROM saved_views WHERE id = ?", id).Scan(
		&existingView.ID,
		&existingView.Name,
		&existingView.ViewType,
		&existingView.ProjectID,
		&existingView.Filters,
		&existingView.SortBy,
		&existingView.SortOrder,
		&existingView.IsDefault,
		&existingView.CreatedAt,
		&existingView.UpdatedAt,
	)

	if err == sql.ErrNoRows {
		panic(middleware.NewNotFoundError("Saved view"))
	}

	if err != nil {
		panic(middleware.NewFetchError("saved view"))
	}

	now := time.Now().Format(time.RFC3339)

	// Unset any existing default for the same project and view_type
	if existingView.ProjectID.Valid {
		database.Exec("UPDATE saved_views SET is_default = 0 WHERE project_id = ? AND view_type = ?", existingView.ProjectID.Int64, existingView.ViewType)
	} else {
		database.Exec("UPDATE saved_views SET is_default = 0 WHERE project_id IS NULL AND view_type = ?", existingView.ViewType)
	}

	// Set this view as default
	_, err = database.Exec("UPDATE saved_views SET is_default = 1, updated_at = ? WHERE id = ?", now, id)
	if err != nil {
		panic(middleware.NewUpdateError("saved view"))
	}

	// Fetch the updated view
	var view SavedView
	err = database.QueryRow(`
		SELECT sv.*, p.name as project_name
		FROM saved_views sv
		LEFT JOIN projects p ON sv.project_id = p.id
		WHERE sv.id = ?
	`, id).Scan(
		&view.ID,
		&view.Name,
		&view.ViewType,
		&view.ProjectID,
		&view.Filters,
		&view.SortBy,
		&view.SortOrder,
		&view.IsDefault,
		&view.CreatedAt,
		&view.UpdatedAt,
		&view.ProjectName,
	)

	if err != nil {
		panic(middleware.NewFetchError("saved view"))
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(parseSavedView(&view)))
}
